pkg/receive/limiter.go

```diff
@@ -5,59 +5,200 @@ package receive
+	writeGate                 gate.Gate
+	registerer                prometheus.Registerer
+	configPathOrContent       fileContent
+	logger                    log.Logger
+	configReloadCounter       prometheus.Counter
+	configReloadFailedCounter prometheus.Counter
+	receiverMode              ReceiverMode
+}
+
+// headSeriesLimiter encompasses active/head series limiting logic.
+type headSeriesLimiter interface {
+	QueryMetaMonitoring(context.Context) error
+	isUnderLimit(tenant string) (bool, error)
 }
 
-// requestLimiter encompasses logic for limiting remote write requests.
 type requestLimiter interface {
 	AllowSizeBytes(tenant string, contentLengthBytes int64) bool
 	AllowSeries(tenant string, amount int64) bool
 	AllowSamples(tenant string, amount int64) bool
 }
 
-// headSeriesLimiter encompasses active/head series limiting logic.
-type headSeriesLimiter interface {
-	QueryMetaMonitoring(context.Context) error
-	isUnderLimit(tenant string) (bool, error)
+// fileContent is an interface to avoid a direct dependency on kingpin or extkingpin.
+type fileContent interface {
+	Content() ([]byte, error)
+	Path() string
 }
 
-func NewLimiter(root *RootLimitsConfig, reg prometheus.Registerer, r ReceiverMode, logger log.Logger) *limiter {
-	limiter := &limiter{
+// NewLimiter creates a new *Limiter given a configuration and prometheus
+// registerer.
+func NewLimiter(configFile fileContent, reg prometheus.Registerer, r ReceiverMode, logger log.Logger) (*Limiter, error) {
+	limiter := &Limiter{
 		writeGate:         gate.NewNoop(),
 		requestLimiter:    &noopRequestLimiter{},
 		HeadSeriesLimiter: NewNopSeriesLimit(),
+		registerer:        reg,
+		logger:            logger,
+		receiverMode:      r,
 	}
-	if root == nil {
-		return limiter
+
+	if reg != nil {
+		limiter.configReloadCounter = promauto.With(reg).NewCounter(
+			prometheus.CounterOpts{
+				Namespace: "thanos",
+				Subsystem: "receive",
+				Name:      "limits_config_reload_total",
+				Help:      "How many times the limit configuration was reloaded",
+			},
+		)
+		limiter.configReloadFailedCounter = promauto.With(reg).NewCounter(
+			prometheus.CounterOpts{
+				Namespace: "thanos",
+				Subsystem: "receive",
+				Name:      "limits_config_reload_err_total",
+				Help:      "How many times the limit configuration failed to reload.",
+			},
+		)
 	}
 
-	maxWriteConcurrency := root.WriteLimits.GlobalLimits.MaxConcurrency
+	if configFile == nil {
+		return limiter, nil
+	}
+
+	limiter.configPathOrContent = configFile
+	if err := limiter.loadConfig(); err != nil {
+		// TODO: wrap error.
+		return nil, err
+	}
+
+	return limiter, nil
+}
+
+// StartConfigReloader starts the automatic configuration reloader based off of
+// the file indicated by pathOrContent. It starts a Go routine in the given
+// *run.Group. Pushes error parsing the reloaded configuration into errChan.
+func (l *Limiter) StartConfigReloader(ctx context.Context, errChan chan<- error) error {
```

Comment:
Is that err channel used actually somewhere, except for the test? Do we need it?

Reply:
It's used only in the tests. I tried to think of something else to check errors without it but I couldn't. Do you have any ideas or suggestions? 🤔

Reply:
I mean if we're not handling the `error`, probably not worth sending it over channel, if we only want to log and move on. Perhaps we should test correctness of`loadConfig()` instead?

Reply:
I don't think that's the case. That test ensures that when reloading, if the configuration in the file is invalid, the reloader stays alive and can successfully reloads again when the configuration is fixed. It means testing the function returned by `PathContentReloader`. 

It only seems like it tests `loadConfig` for correctness because that the only call inside the closure function that can return an error.

Reply:
I'm giving another shot at this.

Reply:
I realized that I don't need to error channel to ensure that the reloader keeps working on this scenario, so I got rid of it and that's all.